File: py/src/imbrace/resources/marketplace.py
from typing import Any, Dict, List, Optional
from ..http import HttpTransport, AsyncHttpTransport
from ..types.marketplace import Product, Order, CreateOrderInput


class MarketplaceResource:
    """Marketplace domain — Sync.

    Two backends:
    - Standalone marketplaces service: gateway/marketplaces/v1
    - Platform v2 marketplaces: gateway/platform/v2/marketplaces

    @param marketplaces_base - standalone marketplaces base (gateway/marketplaces/v1)
    @param platform_base     - platform service base (gateway/platform)
    """

    def __init__(self, http: HttpTransport, marketplaces_base: str, platform_base: str):
        self._http = http
        self._standalone_v1 = f"{marketplaces_base.rstrip('/')}/v1"
        self._platform_v2 = f"{platform_base.rstrip('/')}/v2/marketplaces"

    # --- Standalone marketplaces service ---
    def list_use_case_templates(self) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._standalone_v1}/market-places/templates").json()

    def install_from_json(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._standalone_v1}/market-places/templates/install-from-json", json=body).json()

    # --- Platform v2 products ---
    def list_products(self, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._platform_v2}", params=params or {}).json()

    def get_product(self, product_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._platform_v2}/{product_id}").json()

    def create_product(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._platform_v2}", json=body).json()

    def update_product(self, product_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("PATCH", f"{self._platform_v2}/{product_id}", json=body).json()

    def delete_product(self, product_id: str) -> None:
        self._http.request("DELETE", f"{self._platform_v2}/{product_id}")

    # --- Orders ---
    def create_order(self, body: Dict[str, Any]) -> Dict[str, Any]:
        product_id = body.get("product_id", "")
        return self._http.request("POST", f"{self._platform_v2}/installations/{product_id}", json=body).json()

    def list_orders(self, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._platform_v2}/orders", params=params or {}).json()

    def get_order(self, order_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._platform_v2}/orders/{order_id}").json()

    def update_order_status(self, order_id: str, status: str) -> Dict[str, Any]:
        return self._http.request("PATCH", f"{self._platform_v2}/orders/{order_id}/status", json={"status": status}).json()

    # --- Files ---
    def upload_file(self, files: Any) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._platform_v2}/files", files=files).json()

    def delete_file(self, file_id: str) -> None:
        self._http.request("DELETE", f"{self._platform_v2}/files/{file_id}")

    def get_file_details(self, file_id: str) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._platform_v2}/file-details/{file_id}").json()

    def download_file(self, short_path: str) -> Any:
        return self._http.request("GET", f"{self._platform_v2}/download/{short_path}")

    # --- Email Templates ---
    def list_email_templates(self, params: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._platform_v2}/email-templates/search", params=params or {}).json()

    def create_email_template(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._platform_v2}/email-templates", json=body).json()

    def post_channel_workflows(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.request("POST", f"{self._platform_v2}/channel-workflows", json=body).json()

    # --- Categories ---
    def list_categories(self, params: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
        return self._http.request("GET", f"{self._platform_v2}/categories", params=params or {}).json()


class AsyncMarketplaceResource:
    """Marketplace domain — Async."""

    def __init__(self, http: AsyncHttpTransport, marketplaces_base: str, platform_base: str):
        self._http = http
        self._standalone_v1 = f"{marketplaces_base.rstrip('/')}/v1"
        self._platform_v2 = f"{platform_base.rstrip('/')}/v2/marketplaces"

    async def list_use_case_templates(self) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._standalone_v1}/market-places/templates")
        return res.json()

    async def list_products(self, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._platform_v2}", params=params or {})
        return res.json()

    async def get_product(self, product_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._platform_v2}/{product_id}")
        return res.json()

    async def create_product(self, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("POST", f"{self._platform_v2}", json=body)
        return res.json()

    async def update_product(self, product_id: str, body: Dict[str, Any]) -> Dict[str, Any]:
        res = await self._http.request("PATCH", f"{self._platform_v2}/{product_id}", json=body)
        return res.json()

    async def delete_product(self, product_id: str) -> None:
        await self._http.request("DELETE", f"{self._platform_v2}/{product_id}")

    async def create_order(self, body: Dict[str, Any]) -> Dict[str, Any]:
        product_id = body.get("product_id", "")
        res = await self._http.request("POST", f"{self._platform_v2}/installations/{product_id}", json=body)
        return res.json()

    async def list_orders(self, params: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._platform_v2}/orders", params=params or {})
        return res.json()

    async def get_order(self, order_id: str) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._platform_v2}/orders/{order_id}")
        return res.json()

    async def update_order_status(self, order_id: str, status: str) -> Dict[str, Any]:
        res = await self._http.request("PATCH", f"{self._platform_v2}/orders/{order_id}/status", json={"status": status})
        return res.json()

    async def list_categories(self, params: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
        res = await self._http.request("GET", f"{self._platform_v2}/categories", params=params or {})
        return res.json()

File: ts/src/resources/marketplace.ts
import { HttpTransport } from "../http.js"
import type { Product, Order, CreateOrderInput, OrderStatus, PagedResponse } from "../types/index.js"

export class MarketplaceResource {
  /**
   * @param marketplacesBase - Standalone marketplaces service (gateway/marketplaces/v1)
   * @param platformBase     - Platform service base (gateway/platform) for platform/v2/marketplaces
   */
  constructor(
    private readonly http: HttpTransport,
    private readonly marketplacesBase: string,
    private readonly platformBase: string,
  ) {}

  private get standaloneV1() { return `${this.marketplacesBase}/v1` }
  private get platformV2() { return `${this.platformBase}/v2` }

  // ─── Standalone marketplaces service (/marketplaces/v1) ─────────────────────

  async listUseCaseTemplates() {
    return this.http.getFetch()(`${this.standaloneV1}/market-places/templates`, { method: "GET" }).then(r => r.json())
  }

  async installFromJson(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.standaloneV1}/market-places/templates/install-from-json`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  // ─── Platform marketplaces (/platform/v2/marketplaces) ──────────────────────

  async listProducts(params?: {
    page?: number
    limit?: number
    category?: string
    search?: string
  }): Promise<PagedResponse<Product>> {
    const url = new URL(`${this.platformV2}/marketplaces/products`)
    if (params?.page)     url.searchParams.set("page",     String(params.page))
    if (params?.limit)    url.searchParams.set("limit",    String(params.limit))
    if (params?.category) url.searchParams.set("category", params.category)
    if (params?.search)   url.searchParams.set("search",   params.search)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getProduct(productId: string): Promise<Product> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/products/${productId}`, { method: "GET" }).then(r => r.json())
  }

  async installProduct(productId: string): Promise<Order> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/installations/${productId}`, {
      method: "POST",
    }).then(r => r.json())
  }

  async createOrder(body: CreateOrderInput): Promise<Order> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/installations/${body.product_id ?? ""}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async listOrders(params?: { page?: number; limit?: number; status?: OrderStatus }): Promise<PagedResponse<Order>> {
    const url = new URL(`${this.platformV2}/marketplaces/orders`)
    if (params?.page)   url.searchParams.set("page",   String(params.page))
    if (params?.limit)  url.searchParams.set("limit",  String(params.limit))
    if (params?.status) url.searchParams.set("status", params.status)
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async getOrder(orderId: string): Promise<Order> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/orders/${orderId}`, { method: "GET" }).then(r => r.json())
  }

  async updateOrderStatus(orderId: string, status: OrderStatus): Promise<Order> {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/orders/${orderId}/status`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ status }),
    }).then(r => r.json())
  }

  // ─── Files ───────────────────────────────────────────────────────────────────

  async uploadFile(body: FormData) {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/files`, {
      method: "POST",
      body,
    }).then(r => r.json())
  }

  async deleteFile(fileId: string): Promise<void> {
    await this.http.getFetch()(`${this.platformV2}/marketplaces/files/${fileId}`, { method: "DELETE" })
  }

  async getFileDetails(fileId: string) {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/file-details/${fileId}`, { method: "GET" }).then(r => r.json())
  }

  // ─── Email Templates ─────────────────────────────────────────────────────────

  async listEmailTemplates(params?: Record<string, string>) {
    const url = new URL(`${this.platformV2}/marketplaces/email-templates/search`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  async createEmailTemplate(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/email-templates`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async postChannelWorkflows(body: Record<string, unknown>) {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/channel-workflows`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  async downloadMarketPlaceFile(shortPath: string) {
    return this.http.getFetch()(`${this.platformV2}/marketplaces/download/${shortPath}`, { method: "GET" })
  }

  // ─── Categories ──────────────────────────────────────────────────────────────

  async listCategories(params?: Record<string, string>) {
    const url = new URL(`${this.platformV2}/marketplaces/categories`)
    if (params) Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }
}
